modules/mono/glue/collections_glue.cpp

```diff
@@ -0,0 +1,212 @@
+	return GDMonoMarshal::variant_to_mono_object(ptr->operator[](index));
+}
+
+void godot_icall_Array_SetAt(Array *ptr, int index, MonoObject *value) {
+	if (index < 0 || index > ptr->size())
+		mono_raise_exception(mono_get_exception_index_out_of_range());
+	ptr->operator[](index) = GDMonoMarshal::mono_object_to_variant(value);
+}
+
+int godot_icall_Array_Count(Array *ptr) {
+	return ptr->size();
+}
+
+void godot_icall_Array_Add(Array *ptr, MonoObject *item) {
+	ptr->append(GDMonoMarshal::mono_object_to_variant(item));
+}
+
+void godot_icall_Array_Clear(Array *ptr) {
+	ptr->clear();
+}
+
+bool godot_icall_Array_Contains(Array *ptr, MonoObject *item) {
+	return ptr->find(GDMonoMarshal::mono_object_to_variant(item)) != -1;
+}
+
+int godot_icall_Array_IndexOf(Array *ptr, MonoObject *item) {
+	return ptr->find(GDMonoMarshal::mono_object_to_variant(item));
+}
+
+void godot_icall_Array_Insert(Array *ptr, int index, MonoObject *item) {
+	if (index < 0 || index > ptr->size())
+		mono_raise_exception(mono_get_exception_index_out_of_range());
+	ptr->insert(index, GDMonoMarshal::mono_object_to_variant(item));
+}
+
+bool godot_icall_Array_Remove(Array *ptr, MonoObject *item) {
+	int idx = ptr->find(GDMonoMarshal::mono_object_to_variant(item));
+	if (idx >= 0) {
+		ptr->remove(idx);
+		return true;
+	}
+	return false;
+}
+
+void godot_icall_Array_RemoveAt(Array *ptr, int index) {
+	if (index < 0 || index > ptr->size())
+		mono_raise_exception(mono_get_exception_index_out_of_range());
+	ptr->remove(index);
+}
+
+Dictionary *godot_icall_Dictionary_Ctor() {
+	return memnew(Dictionary);
+}
+
+void godot_icall_Dictionary_Dtor(Dictionary *ptr) {
+	memdelete(ptr);
+}
+
+MonoObject *godot_icall_Dictionary_GetValue(Dictionary *ptr, MonoObject *key) {
+	Variant *ret = ptr->getptr(GDMonoMarshal::mono_object_to_variant(key));
+	if (ret == NULL) {
+		MonoObject *exc = mono_object_new(SCRIPTS_DOMAIN, CACHED_CLASS(KeyNotFoundException)->get_mono_ptr());
+#ifdef DEBUG_ENABLED
+		CRASH_COND(!exc);
+#endif
+		mono_runtime_object_init(exc);
+		mono_raise_exception((MonoException *)exc);
+	}
+	return GDMonoMarshal::variant_to_mono_object(ret);
+}
+
+void godot_icall_Dictionary_SetValue(Dictionary *ptr, MonoObject *key, MonoObject *value) {
+	ptr->operator[](GDMonoMarshal::mono_object_to_variant(key)) = GDMonoMarshal::mono_object_to_variant(value);
+}
+
+Array *godot_icall_Dictionary_Keys(Dictionary *ptr) {
+	return memnew(Array(ptr->keys()));
+}
+
+Array *godot_icall_Dictionary_Values(Dictionary *ptr) {
+	return memnew(Array(ptr->values()));
+}
+
+int godot_icall_Dictionary_Count(Dictionary *ptr) {
+	return ptr->size();
+}
+
+void godot_icall_Dictionary_Add(Dictionary *ptr, MonoObject *key, MonoObject *value) {
+	Variant varKey = GDMonoMarshal::mono_object_to_variant(key);
+	Variant *ret = ptr->getptr(varKey);
+	if (ret != NULL)
+		mono_raise_exception(mono_get_exception_argument("key", "An element with the same key already exists"));
```

Comment:
I haven't looked at godot general implementation of mono embedding, but you need to be careful using mono_raise_exception as it is not going to finish the current stack frame, so in this case it would cause 'varKey' destructor not to be called

Reply:
Thanks. Indeed, that would have messed things up. Now it's using pending exceptions instead.